### DISCUSSION TITLE
Add Header Override / Extends Frontmatter Prop

### DISCUSSION THREAD

**Discussion by Bwc9876** [withastro/starlight#951] [Feature Requests]
### What version of `starlight` are you using?

0.6.1

### What is your idea?

Although auto-generated TOC covers many use cases it can be limiting for people that use custom components, as those headers won't get parsed. So a way to override or perhaps simply extend what headers should be placed in the TOC and their levels, ids, etc should be exposed to the user.

### Why is this feature necessary?

This issue would help users who want to utilize Astro components in their starlight projects more. 

### Do you have examples of this feature in other projects?

- The old Astro docs template had a way to pass headers as an object to its Page component

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by HiDeoo** [comment]
Thanks for your feedback.

Starting with [Starlight 0.32.0](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.32.0), [route data](https://starlight.astro.build/guides/route-data/) provides a way to read and even update all the information required to render a single page, including the table of contents using the [`toc` property](https://starlight.astro.build/reference/route-data/#toc).

Such data can be customized using a [route data middleware](https://starlight.astro.build/guides/route-data/#how-to-customize-route-data). For example, to add some custom items to the table of contents, this would look like this:

```js
// src/routeData.ts
import { defineRouteMiddleware } from "@astrojs/starlight/route-data";

export const onRequest = defineRouteMiddleware((context) => {
  const { starlightRoute } = context.locals;
  starlightRoute.toc = starlightRoute.toc
    ? {
        ...starlightRoute.toc,
        items: [
          ...starlightRoute.toc.items,
          // Add some custom TOC items
          {
            depth: 2,
            slug: "custom-heading",
            text: "Custom heading",
            children: [
              {
                depth: 3,
                slug: "custom-sub-heading",
                text: "Custom sub heading",
                children: [],
              },
            ],
          },
        ],
      }
    : undefined;
});
```

Of course, this can be done more dynamically, or only in some pages, etc.

For [custom pages](https://starlight.astro.build/guides/pages/#custom-pages), the same approach can also be used but the [`<StarlightPage>` component](https://starlight.astro.build/guides/pages/#starlightpage-component) also accepts a [`headings` property](https://starlight.astro.build/guides/pages/#headings) to specify the headings to be used in the table of contents:

```astro
---
// src/pages/custom-page/example.astro
import StarlightPage from '@astrojs/starlight/components/StarlightPage.astro';
---

<StarlightPage frontmatter={{ title: 'My custom page' }} headings={[
  {
    depth: 2,
    slug: "custom-heading",
    text: "Custom heading",
  },
  {
    depth: 3,
    slug: "custom-sub-heading",
    text: "Custom sub heading",
  }
]}>
  <!-- Custom page content -->
</StarlightPage>
```
